### ISSUE TITLE
PerInstancePreferenceModel Documentation

### ISSUE THREAD

**Issue by mcspud** [EliotBerriot/django-dynamic-preferences#56]
Thanks for the great utility library, we have moved all of our settings into it!

We are having trouble when subclassing PerInstancePreferenceRegistry, adding a suite of settings to specific model instances.  This includes creating a registry and adding default settings for each instance.

If someone could point me in the right direction (or give a basic example) I'd be happy to write the documentation and add it as a pull request.

Cheers

**Comment by EliotBerriot** [comment]
Indeed, there is pretty much no documentation regarding this. As a starter, you can have a look at the `UserPreferenceModel` (in `models.py`, which works using this system).

The steps involved are:

1. [Subclassing the `PerInstancePreferenceModel`](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/models.py#L106), to reference the model you want to bind preferences on
2. [Subclassing the `PerInstancePreferenceRegistry`](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/registries.py#L169) (You'll also have to create [an actual instance](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/registries.py#L173) of this registry)
3. [Bind your preference model to your preference registry](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/apps.py#L15). You can do this using a django `AppConfig`, have a look at `apps.py`, basically, you just need a single line: `preference_models.register(PreferenceModel, registry)`. Your app will have to be installed before `dynamic_preferences` in `INSTALLED_APPS` for this to work.
4. Of course, you'll have to generate a migration for your newly created model

After this, you should be able to start registering preferences using `@your_registry_instance.register` decorator.

I'm sorry for the lack of documentation, I'll leave this issue open as a reminder to write more thorough documentation about this specific use case. Hopefully, I'll have some time to do it this week.

In the meantime, feel free to ask any question regarding this process here, in the end, it will be helpful to document tricky steps ;)

**Comment by mcspud** [comment]
Fantastic, thanks.
I'll submit a documentation pull this week.

**Comment by mcspud** [comment]
So I must be missing something here.
The steps you outlined above were (essentially) what we had originally done, but we still couldn't get the preferences to automatically register.

If I check the global preferences registry, I can see my registry and my single (test) preference there:

```
In [2]: preference_models
Out[2]: 
OrderedDict([(xero_accounting.models.XeroPerInstancePreferences,
              OrderedDict([(None, OrderedDict([('TESTTEST', <xero_accounting.settings.TESTTEST at 0x111cf84e0>)]))])),
             (dynamic_preferences.models.UserPreferenceModel, OrderedDict()),
             ...
```

Unfortunately, I can't get the default preferences to populate.  I can see that this [signal hook](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/models.py#L123) goes and searches for registered models, [returning the PersistingTheory Registry back](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/registries.py#L44).

I tried to manually force a data migration to run, whilst adding a more helpful message [exception](https://github.com/EliotBerriot/django-dynamic-preferences/blob/master/dynamic_preferences/registries.py#L54):
```
for contact in Contacts.objects.all():
    XeroPerInstancePreferences.objects.create(name='TESTTEST', raw_value='hello', section=invoice_template_section, instance=contact)
```

Giving: `GlobalPreferenceModel has no field named 'instance'` (but actually creating the DB settings)
The registry also isn't finding these settings
```
In [9]: xero_preferences_registry.manager().all()
Out[9]: {'INVOICE_TEMPLATES__TESTTEST': 'hello'}
```

So it looks like I haven't linked the PerInstance model up to the registry, but I can't see where I've gone wrong:

models.py:
```
@reversion.register(format='json')
class XeroPerInstancePreferences(PerInstancePreferenceModel):

    instance = models.ForeignKey('xero_accounting.Contacts',)

    class Meta(PerInstancePreferenceModel.Meta):
        app_label = 'xero_accounting'
        verbose_name = 'Xero Account Setting'
        verbose_name_plural = '%ss' % verbose_name
```

settings.py
```
@xero_preferences_registry.register
class TESTTEST(StringPreference):
    name = 'TESTTEST'
    default = 'hello'
    verbose_name = '{section}: {name}'.format(section=section, name=name)
    section = invoice_template_section
```

registries.py
```
from dynamic_preferences.registries import PerInstancePreferenceRegistry


class XeroPerInstanceRegistry(PerInstancePreferenceRegistry):
    pass


xero_preferences_registry = XeroPerInstanceRegistry()
```

apps.py
```
from django.apps import AppConfig
from dynamic_preferences.registries import preference_models
from xero_accounting.registries import xero_preferences_registry

class XeroAccountingConfig(AppConfig):
    name = 'xero_accounting'
    verbose_name = 'Xero Accounting'

    def ready(self):
        from xero_accounting import signals
        XeroPerInstancePreferences = self.get_model('XeroPerInstancePreferences')
        preference_models.register(XeroPerInstancePreferences, xero_preferences_registry)
```

**Comment by EliotBerriot** [comment]
Okay, thank you for the details. Here are possible ideas to better understand the problem.

First of all, from the output you gave, your PreferenceModel / registry binding seems okay.

Now, regarding the default preferences not being created:

1. If you're speaking about *existing* `Contact` instances, this is by design (although not documented). `dynamic_preferences` will only populate default preferences for newly created instances
2. If you're speaking about newly created `Contact` instances, then you may have hit a bug.

In any case, can you run the following in a django shell ?

    from xero_accounting.models import Contacts
    from yourpreferences_app.models import XeroPerInstancePreferences
    # We get a random object just to try
    c = Contacts.objects.order_by('?').first()

    # This should try to grab the preferences, fail and 
    # fallback to creating the default ones
    print(c.preferences.all())
    >>> {'TESTTEST': 'hello'}

    # try to change the value
    c.preferences['TESTTEST'] = 'world'

    # we get the updated preferences from db:

    p = XeroPerInstancePreferences.objects.get(
        instance=c,
        code='TESTTEST'
    )

    assert p.value == 'world'


If the previous example works, this means things are basically working,
and you only have a problem with the data migration (this would not
be a surprise, django migrations are sometimes scary).

Also, in your message, you say:

> Giving: GlobalPreferenceModel has no field named 'instance' (but actually creating the DB settings)

Can you please provide the full stacktrace / code generating the stacktrace ?
